{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class biochem_features():\n",
    "    '''returns dataframes for biochemical changes and MAPP score'''\n",
    "    \n",
    "    def __init__(self, *args):\n",
    "        '''housekeeping'''\n",
    "        \n",
    "        self.gene = ['A', 'B', 'C']\n",
    "        \n",
    "        self.aa_list = ['A','C','D','E','F','G','H','I','K','L',\n",
    "                       'M','N','P','Q','R','S','T','V','W','Y']\n",
    "        \n",
    "        self.aa_volumes = {'A': 88.6, 'R': 173.4, 'N': 114.1, 'D': 111.1, 'C': 108.5,\n",
    "                           'Q': 143.8, 'E': 138.4, 'G': 60.1, 'H': 153.2, 'I': 166.7,\n",
    "                           'L': 166.7, 'K': 168.6, 'M': 162.9, 'F': 189.9, 'P': 112.7,\n",
    "                           'S': 89.0, 'T': 116.1, 'W': 227.8, 'Y': 193.6, 'V': 140.0}\n",
    "        \n",
    "        self.aa_MW = {'A': 89.1, 'R': 174.2, 'N': 132.1, 'D': 133.1, 'C': 121.1,\n",
    "                      'E': 147.1, 'Q': 146.2, 'G':75.1, 'H': 155.2, 'I': 131.2,\n",
    "                      'L': 131.2, 'K': 146.2, 'M': 149.2, 'F': 165.2, 'P': 115.1,\n",
    "                      'S': 105.1, 'T': 119.1, 'W': 204.2, 'Y': 181.2, 'V': 117.1}\n",
    "        \n",
    "        self.hydropathy_index = {'A': 1.8, 'R': -4.5, 'N': -3.5, 'D': -3.5,\n",
    "                                 'C': 2.5, 'E': -3.5, 'Q': -3.5, 'G': -0.4,\n",
    "                                 'H': -3.2, 'I': 4.5, 'L': 3.8, 'K': -3.9,\n",
    "                                 'M': 1.9, 'F': 2.8, 'P': -1.6, 'S': -0.8, \n",
    "                                 'T': -0.7, 'W': -0.9, 'Y': -1.3, 'V': 4.2}\n",
    "        \n",
    "        self.Pi = {'A': 6, 'R': 10.76, 'N': 5.41, 'D': 2.77, 'C': 5.07, 'E': 3.22,\n",
    "                   'Q': 5.65, 'G': 5.97, 'H': 7.59, 'I': 6.02, 'L': 5.98, 'K': 9.74,\n",
    "                   'M': 5.74, 'F': 5.48, 'P': 6.3, 'S': 5.68, 'T': 5.6, 'W': 5.89,\n",
    "                   'Y': 5.66, 'V': 5.96}\n",
    "    \n",
    "    \n",
    "    def biochem_change(self):\n",
    "        '''calculates change in biochemical feature for every possible mutation'''\n",
    "        \n",
    "        #create df from attribute function data\n",
    "        df = pd.DataFrame.from_dict([self.aa_volumes, self.aa_MW, \n",
    "                                     self.hydropathy_index, self.Pi]).T.reset_index(0)\n",
    "        DF = df.rename(columns = {'index':'Amino_acid', 0:'volume', 1:'MW', \n",
    "                                  2:'hydropathy',3:'Pi'})\n",
    "        \n",
    "        #create dicts for delta attributes from df\n",
    "        d_volume, d_Pi, d_MW, d_hydropathy = {},{},{},{}\n",
    "        for i in DF.index:\n",
    "            for j in DF.index:\n",
    "                mutation = DF['Amino_acid'][i] + DF['Amino_acid'][j]\n",
    "                d_volume[mutation] = DF['volume'][i] - DF['volume'][j]\n",
    "                d_hydropathy[mutation] = DF['hydropathy'][i] - DF['hydropathy'][j]\n",
    "                d_Pi[mutation] = DF['Pi'][i] - DF['Pi'][j]\n",
    "                d_MW[mutation] = DF['MW'][i] - DF['MW'][j]\n",
    "                \n",
    "        #creates df from delta dictionaries\n",
    "        d_chem_df = pd.DataFrame.from_dict([d_volume, d_MW, \n",
    "                                            d_hydropathy, d_Pi]).T.reset_index(0)\n",
    "        d_chem_DF = d_chem_df.rename(columns={'index':'dAA', 0:'d_volume', \n",
    "                                              1:'d_MW', 2:'d_hydropathy',3:'Pi'})\n",
    "        \n",
    "        return d_chem_DF\n",
    "    \n",
    "    \n",
    "    def MAPP(self):\n",
    "        '''generates df of MAPP scores for every aa combination'''\n",
    "        \n",
    "        #pulls data from each fasta and MAPP output files for each gene and\n",
    "        #stores in dicts/df\n",
    "        record_dict, file_dict = {},{}\n",
    "        for i in self.gene:\n",
    "            record_dict['records{0}'.format(i)] = (\n",
    "                list(SeqIO.parse('../intermediate_data_files/MAPP/Emb{0}_align.fasta'.\n",
    "                                 format(i), 'fasta')))\n",
    "            file_dict['Emb{0}'.format(i), 'MAPP'] = (\n",
    "                pd.read_excel('../intermediate_data_files/MAPP/Emb{0}_output.xlsx'.\n",
    "                              format(i)))\n",
    "            \n",
    "        #creates df for each gene containing MAPP for every possible mutation for the first\n",
    "        #sequence in the fasta alignment file\n",
    "        gene_list = []\n",
    "        for i in self.gene:\n",
    "            Dict, count = {}, 0\n",
    "            sequence = record_dict['records{0}'.format(i)][0].seq\n",
    "            output = file_dict['Emb{0}'.format(i), 'MAPP']\n",
    "            for g in range(0, len(sequence)):\n",
    "                chcter = sequence[g]\n",
    "                if chcter != '-':\n",
    "                    count += 1\n",
    "                    for j in self.aa_list:\n",
    "                        Dict[chcter+str(count)+j] = output[j][g]\n",
    "                        gene_list.append('emb{0}'.format(i))\n",
    "                        \n",
    "            if i == 'A':\n",
    "                df_A = pd.DataFrame.from_dict(Dict, orient='index').reset_index()\n",
    "            elif i == 'B':\n",
    "                df_B = pd.DataFrame.from_dict(Dict, orient='index').reset_index()\n",
    "            elif i == 'C':\n",
    "                df_C = pd.DataFrame.from_dict(Dict, orient='index').reset_index()\n",
    "            \n",
    "        #adds each df to df_A, then creates master df from df_A\n",
    "        df_A = df_A.append(df_B.append(df_C, ignore_index=True), ignore_index=True)\n",
    "        df_A['GENE'] = gene_list\n",
    "        df = df_A.rename(columns={'index':'MUTATION', 0:'MAPP'})\n",
    "        \n",
    "        return df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
